### ISSUE TITLE
feature: query builder

### ISSUE THREAD

**Issue by plandem** [EvidentSolutions/dalesbred#5]
It would be nice to have basic implementation of QueryBuilder. I know it's not difficult to implement it, but it would be nice to have one from box. Like JOOQ. I'm not talking about fully type safe and etc, just provide chaining builder. Something like this:

```
select().from("aaa").where("id = ?");
select("id, name").from("aaa").sort("id");
```

**Comment by komu** [comment]
Sorry, this is not going to happen.

Over the years I've written several query-builders and there are lot of conflicting design dimensions without a clear "right way" that would be suitable for most scenarios. I'd like to see a nice library that would allow programmatic building and manipulation of SQL AST, but it would really need to be at the level of abstract syntax, instead of trying to mimic SQL surface syntax with a "fluent API". I dislike fluent chaining builders, because those don't tend to compose well.

Anyway, I see it as a possibility that I might attempt such a library one day, but it would not be part of Dalesbred, but rather something built on top of Dalesbred. The problems of building SQL dynamically are almost completely orthogonal from the problems that Dalesbred addresses and I don't see the advantage in tying them together. Not at least before the SQL building has been tried in different context and matured a bit.

Anyway, I can really recommend IntelliJ IDEA + Dalesbred plugin because it can offer schema-based syntax highlighting and validation. So you get benefits of static-typing and nice IDE support for SQL without any overhead. So right now I have the following in my editor:

![SQL-editing](https://cloud.githubusercontent.com/assets/3473/7452536/5096f934-f267-11e4-9f8d-f834bee80a35.png)

And that's not just syntax-highlighted, but IDEA actually checks that tables and columns are valid, can navigate through them etc. And Dalesbred IDEA plugin analyzes the SQL to check that ProjectSummary can be instantiated with the returned values.

So the key proposition of Dalesbred is SQL is actually a fine way to write queries and you have tools that enable that to be painless.

So to sum this up, I wouldn't hold my breath waiting for this, but you can always write that builder on top of Dalesbred yourself and publicize it as a library. I could point to the library in the reference documentation.
